src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -5815,6 +5815,219 @@ static QVariant fcnFormatDate( const QVariantList &values, const QgsExpressionCo
   return locale.toString( datetime, format );
 }
 
+static QVariant fcnFormatInterval( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  const QgsInterval interval = QgsExpressionUtils::getInterval( values.at( 0 ), parent, true );
+  QString format = QgsExpressionUtils::getStringValue( values.at( 1 ), parent );
+
+  qlonglong outstanding = interval.seconds() * 1000;
+
+  bool has_Y = format.contains( "%Y" ) || format.contains( "%0Y" )
+               || format.contains( "%00Y" ) || format.contains( "%000Y" );
+  bool has_M = format.contains( "%M" ) || format.contains( "%0M" );
+  bool has_W = format.contains( "%W" ) || format.contains( "%0W" );
+  bool has_D = format.contains( "%D" ) || format.contains( "%0D" );
+  bool has_h = format.contains( "%h" ) || format.contains( "%0h" );
+  bool has_m = format.contains( "%m" ) || format.contains( "%0m" );
+  bool has_s = format.contains( "%s" ) || format.contains( "%0s" );
+  bool has_z = format.contains( "%z" ) || format.contains( "%0z" )
+               || format.contains( "%00z" );
+
+  // Work out which field is the least significant, so we know which one to
+  // round.
+  bool isLeastSignificant_Y = has_Y && !has_M && !has_W && !has_D && !has_h && !has_m && !has_s && !has_z;
+  bool isLeastSignificant_M =           has_M && !has_W && !has_D && !has_h && !has_m && !has_s && !has_z;
+  bool isLeastSignificant_W =                     has_W && !has_D && !has_h && !has_m && !has_s && !has_z;
+  bool isLeastSignificant_D =                               has_D && !has_h && !has_m && !has_s && !has_z;
+  bool isLeastSignificant_h =                                         has_h && !has_m && !has_s && !has_z;
+  bool isLeastSignificant_m =                                                   has_m && !has_s && !has_z;
+  bool isLeastSignificant_s =                                                             has_s && !has_z;
+
+  const qlonglong ms_s = QgsUnitTypes::fromUnitToUnitFactor( Qgis::TemporalUnit::Seconds, Qgis::TemporalUnit::Milliseconds );
+  const qlonglong ms_m = QgsUnitTypes::fromUnitToUnitFactor( Qgis::TemporalUnit::Minutes, Qgis::TemporalUnit::Milliseconds );
+  const qlonglong ms_h = QgsUnitTypes::fromUnitToUnitFactor( Qgis::TemporalUnit::Hours, Qgis::TemporalUnit::Milliseconds );
+  const qlonglong ms_D = QgsUnitTypes::fromUnitToUnitFactor( Qgis::TemporalUnit::Days, Qgis::TemporalUnit::Milliseconds );
+  const qlonglong ms_W = QgsUnitTypes::fromUnitToUnitFactor( Qgis::TemporalUnit::Weeks, Qgis::TemporalUnit::Milliseconds );
+  const qlonglong ms_M = QgsUnitTypes::fromUnitToUnitFactor( Qgis::TemporalUnit::Months, Qgis::TemporalUnit::Milliseconds );
+  const qlonglong ms_Y = QgsUnitTypes::fromUnitToUnitFactor( Qgis::TemporalUnit::Years, Qgis::TemporalUnit::Milliseconds );
```

Comment:
These are all static constants, so we can optimise this a bit:

```suggestion
  constexpr qlonglong ms_s = 1000;
  constexpr qlonglong ms_m = 60000;
  constexpr qlonglong ms_h = 3600000;
  constexpr qlonglong ms_D = 86400000;
  constexpr qlonglong ms_W = 604800000;
  constexpr qlonglong ms_M = 2592000000;
  constexpr qlonglong ms_Y = 31557600000;
```